### ISSUE TITLE
Adapt or conform line endings

### ISSUE THREAD

**Issue by rgbkrk** [jupyter/jupyter_client#204]
We ran into an issue on Hydrogen + IRKernel on Windows with CRLF line endings being sent:

https://github.com/nteract/hydrogen/issues/312#issuecomment-249016058

In my myopic linux centric world, I'd prefer to conform to LF. Where does this adaptation need to happen - at the frontend who is sending the text or at the kernel?

I lean towards fixing it in the frontend. For Hydrogen this means ensuring LF even if the file is CRLF (and allowing them to keep it that way in Atom if they _really_ want that).

/cc @lgeiger

**Comment by rgbkrk** [comment]
The follow on action here is to document what the expectation is for line feeds in the documentation/spec.

**Comment by takluyver** [comment]
As there are more kernel implementations in more languages than there are frontend implementations, I'd also be inclined to say that the frontend should normalise to LF.

**Comment by rgbkrk** [comment]
And we'd assume the same for response messages (payload protocol for instance with %load).

**Comment by takluyver** [comment]
"assume the same" = up to the frontend, or up to the side sending the message?

**Comment by rgbkrk** [comment]
For anything that fills out editor text, we'd have to assume a similar standard (going to receive LF). Or do we?

**Comment by takluyver** [comment]
I could go either way. It seems neater to say that strings in messages should always be normalised to Unix-style line endings (LF) before transmission, but more pragmatic to say that the frontend side is responsible for normalising in both directions, since that's less to fix.

@minrk, any preference?

**Comment by minrk** [comment]
I'd probably say that it's out-of-scope for the protocol itself to enforce that line endings are normalized, but I would at least *encourage* any frontends to normalize input to LF. I'd rather stop the spec at 'unicode code points', and leave it to frontends and kernels to determine what is valid code (I don't know if I would consider it a bug that IRKernel can't handle CRLF, but I would if IPython can't, for instance).

You want to be careful to **not** do normalization like this on *output*, though, because CR is meaningful on its own, and a stdout CR message followed by a stdout LF message should ultimately end up handled the same as CRLF, while CRxyzLF is a different thing. With stream messages in particular, you are not guaranteed that a message includes the complete string (until output is 'done'). In early notebook formats we stripped newlines, effectively enforcing LF, but ran into a mess with the nuances of CR, and now the format includes the output that came from the kernel exactly, which has behaved much more reliably. It's appropriate to do normalization between the message and the display, but none should be done between the message and nbformat, and it shouldn't be assumed that the kernel will do anything about it.

**Comment by takluyver** [comment]
I think it's generally safe to normalise CRLF->LF (rather than just CR->), but maybe there's a case I've overlooked.

**Comment by minrk** [comment]
Yes, I think that's true. From both sides, I think it's reasonable to expect the other side to understand line endings. In particular, I think it should be fine for a frontend or script to open a file and send it exactly to execute_request. Similarly, it should be fine for a kernel to `cat <file>`. We cannot force kernels to normalize output to LF, because it's technically valid to send stream output one codepoint at a time, so a CR might be sent before it knows that an LF will follow. So frontends **must** handle CRLF, at least in stream output. For the same reason, I think kernels **should** handle CRLF on input because I wouldn't want to require someone writing a simple `run-with-jupyter <file>` script to have to deal with CRLF normalization. That is, unless a CRLF-terminated file is invalid input to the language, in which case an error could reasonably be considered the correct result. If R can execute CRLF scripts, then I would consider it a bug in IRKernel that CRLF strings cannot be executed.

At the same time, I think rich frontends like notebooks, nteract, etc. **should** normalize their own content, for simplicity. Hydrogen's in a middle place, where it's dealing with snippets from plain files, not text that it has full control over. If those files are valid for the kernel's language, I would expect them to run in the language's kernel without modification.
